controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -310,8 +310,54 @@ func (r *StorageClassClaimReconciler) reconcileConsumerPhases() (reconcile.Resul
 		// Update the StorageClassClaim status.
 		r.storageClassClaim.Status.Phase = v1alpha1.StorageClassClaimReady
 
+	} else {
+
+		// Reconciliation needed if the StorageClassClaim is present.
+		if err = r.get(r.storageClassClaim); !errors.IsNotFound(err) {
+
+			// Deletion phase.
+			// Update the StorageClassClaim status.
+			r.storageClassClaim.Status.Phase = v1alpha1.StorageClassClaimDeleting
+
+			// Delete StorageClass.
+			existingStorageClass := &storagev1.StorageClass{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: r.storageClassClaim.Name,
+				},
+			}
+			if err = r.get(existingStorageClass); err == nil {
+
+				// Remove finalizer from StorageClass.
+				existingStorageClass.SetFinalizers(remove(existingStorageClass.Finalizers, StorageClassFinalizer))
+				if err := r.update(existingStorageClass); err != nil {
+					return reconcile.Result{}, fmt.Errorf("failed to remove finalizer from StorageClass: %s", err)
+				}
+				if err = r.delete(existingStorageClass); err != nil {
+					return reconcile.Result{}, fmt.Errorf("failed to delete StorageClass %s: %s", existingStorageClass.Name, err)
+				}
+			} else if errors.IsNotFound(err) {
+				r.log.Info("StorageClass already deleted.")
+			} else {
+				return reconcile.Result{}, fmt.Errorf("failed to get StorageClass %s: %s", existingStorageClass.Name, err)
+			}
+
+			// Call `RevokeStorageClassClaim` service on the provider server with StorageClassClaim as a request message.
+			// Check if StorageClassClaim is still exists (it might have been manually removed during the StorageClass
+			// removal above).
+			if !errors.IsNotFound(r.get(r.storageClassClaim)) {
+				_, err = providerClient.RevokeStorageClassClaim(r.ctx, r.storageCluster.Status.ExternalStorage.ConsumerID, r.storageClassClaim.Name)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+			}
+
+			// Remove finalizer from StorageClassClaim on consumer cluster.
+			r.storageClassClaim.SetFinalizers(remove(r.storageClassClaim.GetFinalizers(), StorageClassClaimFinalizer))
+			if err = r.Client.Update(r.ctx, r.storageClassClaim); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to remove finalizer from StorageClassClaim %s: %s", r.storageClassClaim.Name, err)
```

Comment:
Why do we have a finalizer on the claim object? 
A finalizer on an object work against an ownership-based delete as it prevents k8s from sending the delete operation to the dependent objects. 

If the reason for this finalizer is to keep this object alive until all dependent objects are getting removed than it is better to setup `blockOwnerDeletion: true` on the owner ref and use foreground deletion mode when deleting.

Reply:
I've added a `blockOwnerDeletion` field on the dependant resources. However, I've not removed the finalizer logic as of now, because based on [this](https://unofficial-kubernetes.readthedocs.io/en/latest/concepts/workloads/controllers/garbage-collection/), it demands that:

- A finalizer with value "foregroundDeletion" be added to the list of finalizers.
- Passing the `deleteOptions.propagationPolicy` as `foreground`. This is a bit confusing since we don't call `delete` method on SCClaim (and hence can't pass our own `deleteOptions`, since the SCClaim is already in a deletion phase when we fetch it)?

Reply:
Ok, I agree, but for different reasons.

I will start by stating that there is no need to add a finalizer just to make sure dependent resources are getting deleted, this is already taken care of by k8s. But in this case we have special conditions:

1. `storageclass` is NOT owned by `storageclasscalim` as you cannot establish ownership of a clsuter level resource by a namespace level resource
2. we need to ensure that a `revoke` call is dispatched and received on the provider side before we complete the deletion process.

These conditions are a classic use case for the use of finalizers.